File: artifacts/daily-imaan/modules/DailyImaanWidget/ios/DailyImaanWidgetModule.swift
import ExpoModulesCore
import WidgetKit

// Writes ayat data to the shared App Group (group.com.dailyimaan.app) so the
// WidgetKit extension can read it without a network round-trip, then reloads
// all widget timelines so the home screen updates immediately.
//
// `prayerTimesJson` is `{"date":"YYYY-MM-DD","times":{"Fajr":"HH:mm",...}}`
// — the widget uses it to build a multi-entry timeline so the "next prayer"
// line stays correct all day without the app running. Empty string = leave
// whatever was stored (the widget falls back to the legacy pre-formatted
// `widget_next_prayer` string).
public class DailyImaanWidgetModule: Module {
  public func definition() -> ModuleDefinition {
    Name("DailyImaanWidget")

    AsyncFunction("setWidgetData") { (arabic: String, english: String, surahRef: String, nextPrayer: String, prayerTimesJson: String) in
      let defaults = UserDefaults(suiteName: "group.com.dailyimaan.app")
      defaults?.set(arabic, forKey: "widget_arabic")
      defaults?.set(english, forKey: "widget_english")
      defaults?.set(surahRef, forKey: "widget_surah_ref")
      defaults?.set(nextPrayer, forKey: "widget_next_prayer")
      if !prayerTimesJson.isEmpty {
        defaults?.set(prayerTimesJson, forKey: "widget_prayer_times")
      }
      defaults?.synchronize()

      if #available(iOS 14.0, *) {
        WidgetCenter.shared.reloadAllTimelines()
      }
    }
  }
}

File: artifacts/daily-imaan/modules/DailyImaanWidget/android/src/main/java/expo/modules/dailyimaanwidget/DailyImaanWidgetModule.kt
package expo.modules.dailyimaanwidget

import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

// Writes ayat payload to SharedPreferences (daily_imaan_widget) and
// broadcasts an ACTION_APPWIDGET_UPDATE so the home screen widget refreshes.
class DailyImaanWidgetModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("DailyImaanWidget")

    // prayerTimesJson: {"date":"YYYY-MM-DD","times":{"Fajr":"HH:mm",...}} —
    // stored for parity with iOS; the Android widget doesn't render a
    // prayer line yet but the data is there when it does.
    AsyncFunction("setWidgetData") { arabic: String, english: String, surahRef: String, nextPrayer: String, prayerTimesJson: String ->
      val context: Context = appContext.reactContext ?: return@AsyncFunction
      val prefs = context.getSharedPreferences("daily_imaan_widget", Context.MODE_PRIVATE)
      val edit = prefs.edit()
        .putString("widget_arabic", arabic)
        .putString("widget_english", english)
        .putString("widget_surah_ref", surahRef)
        .putString("widget_next_prayer", nextPrayer)
      if (prayerTimesJson.isNotEmpty()) {
        edit.putString("widget_prayer_times", prayerTimesJson)
      }
      edit.apply()

      val mgr = AppWidgetManager.getInstance(context)
      val ids = mgr.getAppWidgetIds(
        ComponentName(context, "com.dailyimaan.DailyAyatAppWidget")
      )
      if (ids.isNotEmpty()) {
        val intent = Intent(AppWidgetManager.ACTION_APPWIDGET_UPDATE).apply {
          putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)
        }
        context.sendBroadcast(intent)
      }
    }
  }
}
